### PR TITLE
DFS refactor work 

### PR DESCRIPTION
**Pull request by yamadapc**
* Using iterative DFS implementation for all cases
* Added unit-tests
---
I need help with:

* running a benchmark

I will do clean-up around:

* safe roll-out (like a feature-flag or more tests)
* remove old method

Missing from this diff is:

* Replacing `getChildren` with a few generator/callback based options

### CODE REVIEW

File: packages/core/graph/src/Graph.js [hunk 1]
```diff
@@ -11,6 +11,7 @@ import type {
 import {BitSet} from './BitSet';
 
 import nullthrows from 'nullthrows';
+import {getFeatureFlag} from '@parcel/feature-flags/src';
```
Comment (JakeLane) [inline packages/core/graph/src/Graph.js]:
```suggestion
import {getFeatureFlag} from '@parcel/feature-flags';
```

File: packages/core/graph/src/Graph.js [hunk 2]
```diff
@@ -289,11 +335,17 @@ export default class Graph<TNode, TEdgeType: number = 1> {
     ) {
       return this.dfsFast(enter, startNodeId);
```
Comment (JakeLane) [inline packages/core/graph/src/Graph.js:336]: Would be nice to have a comment here to explain why there are two different dfs implementations

File: packages/core/graph/src/Graph.js [hunk 3]
```diff
@@ -449,15 +501,117 @@ export default class Graph<TNode, TEdgeType: number = 1> {
...
+          this._visited = visited;
+          return context;
+        }
+      } else {
+        let {nodeId, context} = command;
+        if (!this.hasNode(nodeId) || visited.has(nodeId)) continue;
+        visited.add(nodeId);
+
+        skipped = false;
+        let enter = typeof visit === 'function' ? visit : visit.enter;
```
Comment (JakeLane) [inline packages/core/graph/src/Graph.js]: Can we move this comparison outside of the loop?

File: packages/core/graph/src/Graph.js (hunk 2 above)
Comment (yamadapc) [reply packages/core/graph/src/Graph.js:336]: Added JSDoc comments

File: packages/core/graph/src/Graph.js (hunk 3 above)
Comment (yamadapc) [reply packages/core/graph/src/Graph.js]: Done, thank you

File: packages/core/graph/src/Graph.js [hunk 4]
```diff
@@ -449,15 +501,122 @@ export default class Graph<TNode, TEdgeType: number = 1> {
...
+      {nodeId: traversalStartNode, context: null},
+    ];
+    const enter = typeof visit === 'function' ? visit : visit.enter;
+    while (queue.length !== 0) {
+      const command = queue.pop();
+
+      if (command.exit != null) {
+        let {nodeId, context, exit} = command;
+        let newContext = exit(nodeId, command.context, actions);
+        if (typeof newContext !== 'undefined') {
```
Comment (MonicaOlejniczak) [inline packages/core/graph/src/Graph.js:552]: /nit `!= null` checks are usually preferred in the codebase from what I can tell

**Review by MonicaOlejniczak — APPROVED** [review]

File: packages/core/graph/src/Graph.js [hunk 5]
```diff
@@ -449,15 +501,122 @@ export default class Graph<TNode, TEdgeType: number = 1> {
...
+        if (skipped) {
+          continue;
+        }
+
+        if (stopped) {
+          this._visited = visited;
+          return context;
+        }
+
+        if (typeof visit !== 'function' && visit.exit) {
```
Comment (marcins) [inline packages/core/graph/src/Graph.js:588]: We assign `enter` to a const up on line 545 - any reason we don't do the same for `exit`? I'm guessing a `typeof` isn't a huge overhead anyway..

**Review by marcins — APPROVED** [review]

**Review by JakeLane — APPROVED** [review]

**Comment by devongovett** [comment]
Do you have performance numbers for this? Where did this work come from?

**Pull request merged**